In [1]:
#importing necessary libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
#without encoding, getting utf-8 error.
laptop_data = pd.read_csv('laptops.csv', encoding='latin-1')
laptop_data.head()

Unnamed: 0 Company      Product   TypeName  Inches  \
0           1   Apple  MacBook Pro  Ultrabook    13.3   
1           2   Apple  Macbook Air  Ultrabook    13.3   
2           3      HP       250 G6   Notebook    15.6   
3           4   Apple  MacBook Pro  Ultrabook    15.4   
4           5   Apple  MacBook Pro  Ultrabook    13.3   

                     ScreenResolution                         Cpu   Ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz   8GB   
1                            1440x900        Intel Core i5 1.8GHz   8GB   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz   8GB   
3  IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz  16GB   
4  IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz   8GB   

                Memory                           Gpu  OpSys  Weight  \
0            128GB SSD  Intel Iris Plus Graphics 640  macOS  1.37kg   
1  128GB Flash Storage        Intel HD Graphics 6000  macOS  1.34kg   
2            256GB SSD         Intel HD Graphics 620  No OS  1.86kg   
3            512GB SSD            AMD Radeon Pro 455  macOS  1.83kg   
4            256GB SSD  Intel Iris Plus Graphics 650  macOS  1.37kg   

   Price_euros  
0      1339.69  
1       898.94  
2       575.00  
3      2537.45  
4      1803.60

Name of the dataframe itself incates that the data is of laptops. Our data contains specifications of the laptops: Company name, Product name, Type of the product, Screen size in inches, Screen Resolution, CPU, System Memory (RAM), Storage Type (Memory), Graphics Processing Unit(GPU), Operating System(opSys), Weight of the product, Product Weight (in kg) and price (in euros).

In [3]:
laptop_data.shape

(1303, 13)

We don't have much more data; we have just 1303 rows and 13 columns. More data can make our prediction better.

In [4]:
print(laptop_data['Company'].unique())

['Apple' 'HP' 'Acer' 'Asus' 'Dell' 'Lenovo' 'Chuwi' 'MSI' 'Microsoft'
 'Toshiba' 'Huawei' 'Xiaomi' 'Vero' 'Razer' 'Mediacom' 'Samsung' 'Google'
 'Fujitsu' 'LG']


In [5]:
print(laptop_data['Company'].value_counts())

Dell         297
Lenovo       297
HP           274
Asus         158
Acer         103
MSI           54
Toshiba       48
Apple         21
Samsung        9
Razer          7
Mediacom       7
Microsoft      6
Xiaomi         4
Vero           4
Chuwi          3
Google         3
Fujitsu        3
LG             3
Huawei         2
Name: Company, dtype: int64


We have different 19 company's data.

In [6]:
laptop_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        1303 non-null   int64  
 1   Company           1303 non-null   object 
 2   Product           1303 non-null   object 
 3   TypeName          1303 non-null   object 
 4   Inches            1303 non-null   float64
 5   ScreenResolution  1303 non-null   object 
 6   Cpu               1303 non-null   object 
 7   Ram               1303 non-null   object 
 8   Memory            1303 non-null   object 
 9   Gpu               1303 non-null   object 
 10  OpSys             1303 non-null   object 
 11  Weight            1303 non-null   object 
 12  Price_euros       1303 non-null   float64
dtypes: float64(2), int64(1), object(10)
memory usage: 132.5+ KB


We don't have any missing value. Unnamed is not useful, Ram and weight data type shows object, we need to remove 'GB' and 'kg' to make the datatype float. 

In [7]:
laptop_data.duplicated().sum()

0

we have no duplicate value. If we have duplicate value, we need to use drop_duplicates() to drop the duplicate value.

In [8]:
laptop_data.isnull().sum()

Unnamed: 0          0
Company             0
Product             0
TypeName            0
Inches              0
ScreenResolution    0
Cpu                 0
Ram                 0
Memory              0
Gpu                 0
OpSys               0
Weight              0
Price_euros         0
dtype: int64

No Missing Value

# Removing Unnecessary Columns

In [9]:
laptop_data.drop(columns=['Unnamed: 0'],inplace=True)
laptop_data.sample(10)

Company                             Product            TypeName  Inches  \
1158   Lenovo  B51-80 (i5-6200U/8GB/1008GB/Radeon            Notebook    15.6   
793    Lenovo                      Yoga 510-15IKB  2 in 1 Convertible    15.6   
474      Asus                         ZenBook Pro            Notebook    15.6   
496    Lenovo                  Ideapad 520-15IKBR            Notebook    15.6   
971      Dell                      Latitude E7470           Ultrabook    14.0   
988      Dell                       Inspiron 5567            Notebook    15.6   
764      Dell                       Latitude 3380            Notebook    13.3   
706      Acer                              Spin 5  2 in 1 Convertible    13.3   
973   Toshiba                     Tecra A40-C-1DF            Notebook    14.0   
99         HP                     Omen 15-ce007nv              Gaming    15.6   

                               ScreenResolution                          Cpu  \
1158                          Full HD 1920x1080   Intel Core i5 6200U 2.3GHz   
793             Full HD / Touchscreen 1920x1080   Intel Core i5 7200U 2.5GHz   
474             IPS Panel 4K Ultra HD 3840x2160  Intel Core i7 6700HQ 2.6GHz   
496                           Full HD 1920x1080   Intel Core i7 8550U 1.8GHz   
971                       Touchscreen 2560x1440   Intel Core i7 6600U 2.6GHz   
988                                    1366x768   Intel Core i7 7500U 2.7GHz   
764                                    1366x768     Intel Core i3 6006U 2GHz   
706   IPS Panel Full HD / Touchscreen 1920x1080   Intel Core i5 8250U 1.6GHz   
973                           Full HD 1920x1080   Intel Core i5 6200U 2.3GHz   
99                  IPS Panel Full HD 1920x1080  Intel Core i7 7700HQ 2.8GHz   

       Ram                Memory                      Gpu       OpSys  Weight  \
1158   8GB          1.0TB Hybrid       AMD Radeon R5 M330  Windows 10   2.5kg   
793    8GB             256GB SSD       AMD Radeon R7 M460  Windows 10  2.08kg   
474   12GB  128GB SSD +  1TB HDD    Intel HD Graphics 530  Windows 10  2.06kg   
496    8GB             256GB SSD   Intel UHD Graphics 620  Windows 10  2.17kg   
971    8GB             256GB SSD    Intel HD Graphics 520  Windows 10   1.5kg   
988    8GB               1TB HDD       AMD Radeon R7 M445  Windows 10  2.36kg   
764    4GB             500GB HDD    Intel HD Graphics 520  Windows 10  1.65kg   
706    8GB             256GB SSD   Intel UHD Graphics 620  Windows 10   1.6kg   
973    8GB             256GB SSD    Intel HD Graphics 520  Windows 10  1.95kg   
99    12GB  128GB SSD +  1TB HDD  Nvidia GeForce GTX 1050  Windows 10  2.62kg   

      Price_euros  
1158       788.49  
793        959.00  
474       1299.00  
496        898.90  
971       1962.98  
988        749.00  
764        849.00  
706        999.00  
973       1180.00  
99        1249.00

We removed first column 'Unnamed: 0' permanently by using inplace=True.

# Correcting Spelling Mistakes

In [10]:
laptop_data["TypeName"] = laptop_data['TypeName'].replace('Netbook', 'Notebook')

In [11]:
laptop_data['TypeName'].value_counts()

Notebook              752
Gaming                205
Ultrabook             196
2 in 1 Convertible    121
Workstation            29
Name: TypeName, dtype: int64

# Cleaning 'Ram' and 'Weight' columns

In [12]:
laptop_data['Ram'] = laptop_data['Ram'].str.replace('GB', '')
laptop_data['Weight'] = laptop_data['Weight'].str.replace('kg', ' ')

In [13]:
laptop_data.head()

Company      Product   TypeName  Inches                    ScreenResolution  \
0   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   
1   Apple  Macbook Air  Ultrabook    13.3                            1440x900   
2      HP       250 G6   Notebook    15.6                   Full HD 1920x1080   
3   Apple  MacBook Pro  Ultrabook    15.4  IPS Panel Retina Display 2880x1800   
4   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   

                          Cpu Ram               Memory  \
0        Intel Core i5 2.3GHz   8            128GB SSD   
1        Intel Core i5 1.8GHz   8  128GB Flash Storage   
2  Intel Core i5 7200U 2.5GHz   8            256GB SSD   
3        Intel Core i7 2.7GHz  16            512GB SSD   
4        Intel Core i5 3.1GHz   8            256GB SSD   

                            Gpu  OpSys Weight  Price_euros  
0  Intel Iris Plus Graphics 640  macOS  1.37       1339.69  
1        Intel HD Graphics 6000  macOS  1.34        898.94  
2         Intel HD Graphics 620  No OS  1.86        575.00  
3            AMD Radeon Pro 455  macOS  1.83       2537.45  
4  Intel Iris Plus Graphics 650  macOS  1.37       1803.60

For better prediction, we need to change data in numeric form. We removed 'GB' and 'kg' from Ram and weight. But still the datatype of the columns is object. In order to change the datatype of this two column, we will use astype() and pass int and float respectively.

In [14]:
laptop_data['Ram'] = laptop_data['Ram'].astype('int32')
laptop_data['Weight'] = laptop_data['Weight'].astype('float32')

In [15]:
laptop_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Company           1303 non-null   object 
 1   Product           1303 non-null   object 
 2   TypeName          1303 non-null   object 
 3   Inches            1303 non-null   float64
 4   ScreenResolution  1303 non-null   object 
 5   Cpu               1303 non-null   object 
 6   Ram               1303 non-null   int32  
 7   Memory            1303 non-null   object 
 8   Gpu               1303 non-null   object 
 9   OpSys             1303 non-null   object 
 10  Weight            1303 non-null   float32
 11  Price_euros       1303 non-null   float64
dtypes: float32(1), float64(2), int32(1), object(8)
memory usage: 112.1+ KB


Now, the Ram and Weight column changed to int and float respectively.

# Converting 'weight' and 'price' in desirable unit

In [16]:
laptop_data ['Weight_lb'] = laptop_data ['Weight'] * 2.205
laptop_data['$Price'] = round(laptop_data['Price_euros'] * 1.14,2)
laptop_data.head()

Company      Product   TypeName  Inches                    ScreenResolution  \
0   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   
1   Apple  Macbook Air  Ultrabook    13.3                            1440x900   
2      HP       250 G6   Notebook    15.6                   Full HD 1920x1080   
3   Apple  MacBook Pro  Ultrabook    15.4  IPS Panel Retina Display 2880x1800   
4   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   

                          Cpu  Ram               Memory  \
0        Intel Core i5 2.3GHz    8            128GB SSD   
1        Intel Core i5 1.8GHz    8  128GB Flash Storage   
2  Intel Core i5 7200U 2.5GHz    8            256GB SSD   
3        Intel Core i7 2.7GHz   16            512GB SSD   
4        Intel Core i5 3.1GHz    8            256GB SSD   

                            Gpu  OpSys  Weight  Price_euros  Weight_lb  \
0  Intel Iris Plus Graphics 640  macOS    1.37      1339.69    3.02085   
1        Intel HD Graphics 6000  macOS    1.34       898.94    2.95470   
2         Intel HD Graphics 620  No OS    1.86       575.00    4.10130   
3            AMD Radeon Pro 455  macOS    1.83      2537.45    4.03515   
4  Intel Iris Plus Graphics 650  macOS    1.37      1803.60    3.02085   

    $Price  
0  1527.25  
1  1024.79  
2   655.50  
3  2892.69  
4  2056.10

We are making this model for USA. The pound is a unit of mass used in United States customary systems of measurement. We changed the weight to lb from kg, price to dollar from euro. Saved these columns permanently.

In [17]:
laptop_data.drop(['Weight', 'Price_euros'], axis=1, inplace=True)

# Fetching the information from 'ScreenResolution' column

In [18]:
laptop_data['TouchScreen'] = laptop_data['ScreenResolution'].apply(lambda x:1 if 'Touchscreen' in x else 0)
laptop_data.head()

Company      Product   TypeName  Inches                    ScreenResolution  \
0   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   
1   Apple  Macbook Air  Ultrabook    13.3                            1440x900   
2      HP       250 G6   Notebook    15.6                   Full HD 1920x1080   
3   Apple  MacBook Pro  Ultrabook    15.4  IPS Panel Retina Display 2880x1800   
4   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   

                          Cpu  Ram               Memory  \
0        Intel Core i5 2.3GHz    8            128GB SSD   
1        Intel Core i5 1.8GHz    8  128GB Flash Storage   
2  Intel Core i5 7200U 2.5GHz    8            256GB SSD   
3        Intel Core i7 2.7GHz   16            512GB SSD   
4        Intel Core i5 3.1GHz    8            256GB SSD   

                            Gpu  OpSys  Weight_lb   $Price  TouchScreen  
0  Intel Iris Plus Graphics 640  macOS    3.02085  1527.25            0  
1        Intel HD Graphics 6000  macOS    2.95470  1024.79            0  
2         Intel HD Graphics 620  No OS    4.10130   655.50            0  
3            AMD Radeon Pro 455  macOS    4.03515  2892.69            0  
4  Intel Iris Plus Graphics 650  macOS    3.02085  2056.10            0

In [19]:
laptop_data['IPS'] = laptop_data['ScreenResolution'].apply(lambda x:1 if 'IPS' in x else 0)
laptop_data.head()

Company      Product   TypeName  Inches                    ScreenResolution  \
0   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   
1   Apple  Macbook Air  Ultrabook    13.3                            1440x900   
2      HP       250 G6   Notebook    15.6                   Full HD 1920x1080   
3   Apple  MacBook Pro  Ultrabook    15.4  IPS Panel Retina Display 2880x1800   
4   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   

                          Cpu  Ram               Memory  \
0        Intel Core i5 2.3GHz    8            128GB SSD   
1        Intel Core i5 1.8GHz    8  128GB Flash Storage   
2  Intel Core i5 7200U 2.5GHz    8            256GB SSD   
3        Intel Core i7 2.7GHz   16            512GB SSD   
4        Intel Core i5 3.1GHz    8            256GB SSD   

                            Gpu  OpSys  Weight_lb   $Price  TouchScreen  IPS  
0  Intel Iris Plus Graphics 640  macOS    3.02085  1527.25            0    1  
1        Intel HD Graphics 6000  macOS    2.95470  1024.79            0    0  
2         Intel HD Graphics 620  No OS    4.10130   655.50            0    0  
3            AMD Radeon Pro 455  macOS    4.03515  2892.69            0    1  
4  Intel Iris Plus Graphics 650  macOS    3.02085  2056.10            0    1

In [20]:
laptop_data['ScreenResolution']=laptop_data['ScreenResolution'].apply(lambda x: x.split()[-1])
laptop_data.head()

Company      Product   TypeName  Inches ScreenResolution  \
0   Apple  MacBook Pro  Ultrabook    13.3        2560x1600   
1   Apple  Macbook Air  Ultrabook    13.3         1440x900   
2      HP       250 G6   Notebook    15.6        1920x1080   
3   Apple  MacBook Pro  Ultrabook    15.4        2880x1800   
4   Apple  MacBook Pro  Ultrabook    13.3        2560x1600   

                          Cpu  Ram               Memory  \
0        Intel Core i5 2.3GHz    8            128GB SSD   
1        Intel Core i5 1.8GHz    8  128GB Flash Storage   
2  Intel Core i5 7200U 2.5GHz    8            256GB SSD   
3        Intel Core i7 2.7GHz   16            512GB SSD   
4        Intel Core i5 3.1GHz    8            256GB SSD   

                            Gpu  OpSys  Weight_lb   $Price  TouchScreen  IPS  
0  Intel Iris Plus Graphics 640  macOS    3.02085  1527.25            0    1  
1        Intel HD Graphics 6000  macOS    2.95470  1024.79            0    0  
2         Intel HD Graphics 620  No OS    4.10130   655.50            0    0  
3            AMD Radeon Pro 455  macOS    4.03515  2892.69            0    1  
4  Intel Iris Plus Graphics 650  macOS    3.02085  2056.10            0    1

In [21]:
#laptop_data['ScreenResolution']=laptop_data['ScreenResolution'].str.extract('(\d\d\d\dx\d\d\d\d)')
#laptop_data.head()

ScreenResolution column contains informations like Touchscreen, IPS and resolution of the screen. By using individual lambda function, we fetch the info and saved in diffeent columns. Further, we by using split function, we split x resolution and y resolution.

In [22]:
laptop_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Company           1303 non-null   object 
 1   Product           1303 non-null   object 
 2   TypeName          1303 non-null   object 
 3   Inches            1303 non-null   float64
 4   ScreenResolution  1303 non-null   object 
 5   Cpu               1303 non-null   object 
 6   Ram               1303 non-null   int32  
 7   Memory            1303 non-null   object 
 8   Gpu               1303 non-null   object 
 9   OpSys             1303 non-null   object 
 10  Weight_lb         1303 non-null   float32
 11  $Price            1303 non-null   float64
 12  TouchScreen       1303 non-null   int64  
 13  IPS               1303 non-null   int64  
dtypes: float32(1), float64(2), int32(1), int64(2), object(8)
memory usage: 132.5+ KB


In [23]:
new_df = laptop_data['ScreenResolution'].str.split('x', expand=True)
new_df

0     1
0     2560  1600
1     1440   900
2     1920  1080
3     2880  1800
4     2560  1600
...    ...   ...
1298  1920  1080
1299  3200  1800
1300  1366   768
1301  1366   768
1302  1366   768

[1303 rows x 2 columns]

In [24]:
laptop_data['X_res'] = new_df[0]
laptop_data['Y_res'] = new_df[1]

In [25]:
laptop_data.head()

Company      Product   TypeName  Inches ScreenResolution  \
0   Apple  MacBook Pro  Ultrabook    13.3        2560x1600   
1   Apple  Macbook Air  Ultrabook    13.3         1440x900   
2      HP       250 G6   Notebook    15.6        1920x1080   
3   Apple  MacBook Pro  Ultrabook    15.4        2880x1800   
4   Apple  MacBook Pro  Ultrabook    13.3        2560x1600   

                          Cpu  Ram               Memory  \
0        Intel Core i5 2.3GHz    8            128GB SSD   
1        Intel Core i5 1.8GHz    8  128GB Flash Storage   
2  Intel Core i5 7200U 2.5GHz    8            256GB SSD   
3        Intel Core i7 2.7GHz   16            512GB SSD   
4        Intel Core i5 3.1GHz    8            256GB SSD   

                            Gpu  OpSys  Weight_lb   $Price  TouchScreen  IPS  \
0  Intel Iris Plus Graphics 640  macOS    3.02085  1527.25            0    1   
1        Intel HD Graphics 6000  macOS    2.95470  1024.79            0    0   
2         Intel HD Graphics 620  No OS    4.10130   655.50            0    0   
3            AMD Radeon Pro 455  macOS    4.03515  2892.69            0    1   
4  Intel Iris Plus Graphics 650  macOS    3.02085  2056.10            0    1   

  X_res Y_res  
0  2560  1600  
1  1440   900  
2  1920  1080  
3  2880  1800  
4  2560  1600

In [26]:
laptop_data['X_res']=laptop_data['X_res'].astype('int')
laptop_data['Y_res']=laptop_data['Y_res'].astype('int')
laptop_data.head()

Company      Product   TypeName  Inches ScreenResolution  \
0   Apple  MacBook Pro  Ultrabook    13.3        2560x1600   
1   Apple  Macbook Air  Ultrabook    13.3         1440x900   
2      HP       250 G6   Notebook    15.6        1920x1080   
3   Apple  MacBook Pro  Ultrabook    15.4        2880x1800   
4   Apple  MacBook Pro  Ultrabook    13.3        2560x1600   

                          Cpu  Ram               Memory  \
0        Intel Core i5 2.3GHz    8            128GB SSD   
1        Intel Core i5 1.8GHz    8  128GB Flash Storage   
2  Intel Core i5 7200U 2.5GHz    8            256GB SSD   
3        Intel Core i7 2.7GHz   16            512GB SSD   
4        Intel Core i5 3.1GHz    8            256GB SSD   

                            Gpu  OpSys  Weight_lb   $Price  TouchScreen  IPS  \
0  Intel Iris Plus Graphics 640  macOS    3.02085  1527.25            0    1   
1        Intel HD Graphics 6000  macOS    2.95470  1024.79            0    0   
2         Intel HD Graphics 620  No OS    4.10130   655.50            0    0   
3            AMD Radeon Pro 455  macOS    4.03515  2892.69            0    1   
4  Intel Iris Plus Graphics 650  macOS    3.02085  2056.10            0    1   

   X_res  Y_res  
0   2560   1600  
1   1440    900  
2   1920   1080  
3   2880   1800  
4   2560   1600

In [27]:
laptop_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Company           1303 non-null   object 
 1   Product           1303 non-null   object 
 2   TypeName          1303 non-null   object 
 3   Inches            1303 non-null   float64
 4   ScreenResolution  1303 non-null   object 
 5   Cpu               1303 non-null   object 
 6   Ram               1303 non-null   int32  
 7   Memory            1303 non-null   object 
 8   Gpu               1303 non-null   object 
 9   OpSys             1303 non-null   object 
 10  Weight_lb         1303 non-null   float32
 11  $Price            1303 non-null   float64
 12  TouchScreen       1303 non-null   int64  
 13  IPS               1303 non-null   int64  
 14  X_res             1303 non-null   int32  
 15  Y_res             1303 non-null   int32  
dtypes: float32(1), float64(2), int32(3), int64

# Pixel per Inches (PPI)

In [28]:
laptop_data['ppi'] = (((laptop_data['X_res']**2) + (laptop_data['Y_res']**2))**0.5).astype('float')

In [29]:
laptop_data.corr()['$Price']

Inches         0.068197
Ram            0.743007
Weight_lb      0.210370
$Price         1.000000
TouchScreen    0.191226
IPS            0.252208
X_res          0.556529
Y_res          0.552809
ppi            0.556182
Name: $Price, dtype: float64

In [30]:
laptop_data = laptop_data.drop(columns=['X_res', 'Y_res', 'ScreenResolution'])

# Fetching the information from 'Cpu' column

In [31]:
def processor_type(text):
    if text.split()[0] == 'Intel':
        return 'Intel Processor'
    else:
        return 'AMD Processor'

In [32]:
laptop_data['Cpu processor'] = laptop_data['Cpu'].apply(processor_type)

'Cpu' column contains type of the processor and speed of the processor. By creating a function called process_type, we fetch processor type and creating lambda function we fetch speed of the processor.

In [33]:
laptop_data['Processor_speed_GHz'] = laptop_data['Cpu'].apply(lambda x: x.split()[-1])
laptop_data

Company                              Product            TypeName  Inches  \
0      Apple                          MacBook Pro           Ultrabook    13.3   
1      Apple                          Macbook Air           Ultrabook    13.3   
2         HP                               250 G6            Notebook    15.6   
3      Apple                          MacBook Pro           Ultrabook    15.4   
4      Apple                          MacBook Pro           Ultrabook    13.3   
...      ...                                  ...                 ...     ...   
1298  Lenovo                       Yoga 500-14ISK  2 in 1 Convertible    14.0   
1299  Lenovo                       Yoga 900-13ISK  2 in 1 Convertible    13.3   
1300  Lenovo                   IdeaPad 100S-14IBR            Notebook    14.0   
1301      HP  15-AC110nv (i7-6500U/6GB/1TB/Radeon            Notebook    15.6   
1302    Asus  X553SA-XX031T (N3050/4GB/500GB/W10)            Notebook    15.6   

                                       Cpu  Ram               Memory  \
0                     Intel Core i5 2.3GHz    8            128GB SSD   
1                     Intel Core i5 1.8GHz    8  128GB Flash Storage   
2               Intel Core i5 7200U 2.5GHz    8            256GB SSD   
3                     Intel Core i7 2.7GHz   16            512GB SSD   
4                     Intel Core i5 3.1GHz    8            256GB SSD   
...                                    ...  ...                  ...   
1298            Intel Core i7 6500U 2.5GHz    4            128GB SSD   
1299            Intel Core i7 6500U 2.5GHz   16            512GB SSD   
1300  Intel Celeron Dual Core N3050 1.6GHz    2   64GB Flash Storage   
1301            Intel Core i7 6500U 2.5GHz    6              1TB HDD   
1302  Intel Celeron Dual Core N3050 1.6GHz    4            500GB HDD   

                               Gpu       OpSys  Weight_lb   $Price  \
0     Intel Iris Plus Graphics 640       macOS    3.02085  1527.25   
1           Intel HD Graphics 6000       macOS    2.95470  1024.79   
2            Intel HD Graphics 620       No OS    4.10130   655.50   
3               AMD Radeon Pro 455       macOS    4.03515  2892.69   
4     Intel Iris Plus Graphics 650       macOS    3.02085  2056.10   
...                            ...         ...        ...      ...   
1298         Intel HD Graphics 520  Windows 10    3.96900   727.32   
1299         Intel HD Graphics 520  Windows 10    2.86650  1708.86   
1300             Intel HD Graphics  Windows 10    3.30750   261.06   
1301            AMD Radeon R5 M330  Windows 10    4.82895   870.96   
1302             Intel HD Graphics  Windows 10    4.85100   420.66   

      TouchScreen  IPS          ppi    Cpu processor Processor_speed_GHz  
0               0    1  3018.873962  Intel Processor              2.3GHz  
1               0    0  1698.116604  Intel Processor              1.8GHz  
2               0    0  2202.907170  Intel Processor              2.5GHz  
3               0    1  3396.233208  Intel Processor              2.7GHz  
4               0    1  3018.873962  Intel Processor              3.1GHz  
...           ...  ...          ...              ...                 ...  
1298            1    1  2202.907170  Intel Processor              2.5GHz  
1299            1    1  3671.511950  Intel Processor              2.5GHz  
1300            0    0  1567.092850  Intel Processor              1.6GHz  
1301            0    0  1567.092850  Intel Processor              2.5GHz  
1302            0    0  1567.092850  Intel Processor              1.6GHz  

[1303 rows x 16 columns]

In [34]:
laptop_data['Processor_speed_GHz']=laptop_data['Processor_speed_GHz'].str.replace('GHz', '')
laptop_data

Company                              Product            TypeName  Inches  \
0      Apple                          MacBook Pro           Ultrabook    13.3   
1      Apple                          Macbook Air           Ultrabook    13.3   
2         HP                               250 G6            Notebook    15.6   
3      Apple                          MacBook Pro           Ultrabook    15.4   
4      Apple                          MacBook Pro           Ultrabook    13.3   
...      ...                                  ...                 ...     ...   
1298  Lenovo                       Yoga 500-14ISK  2 in 1 Convertible    14.0   
1299  Lenovo                       Yoga 900-13ISK  2 in 1 Convertible    13.3   
1300  Lenovo                   IdeaPad 100S-14IBR            Notebook    14.0   
1301      HP  15-AC110nv (i7-6500U/6GB/1TB/Radeon            Notebook    15.6   
1302    Asus  X553SA-XX031T (N3050/4GB/500GB/W10)            Notebook    15.6   

                                       Cpu  Ram               Memory  \
0                     Intel Core i5 2.3GHz    8            128GB SSD   
1                     Intel Core i5 1.8GHz    8  128GB Flash Storage   
2               Intel Core i5 7200U 2.5GHz    8            256GB SSD   
3                     Intel Core i7 2.7GHz   16            512GB SSD   
4                     Intel Core i5 3.1GHz    8            256GB SSD   
...                                    ...  ...                  ...   
1298            Intel Core i7 6500U 2.5GHz    4            128GB SSD   
1299            Intel Core i7 6500U 2.5GHz   16            512GB SSD   
1300  Intel Celeron Dual Core N3050 1.6GHz    2   64GB Flash Storage   
1301            Intel Core i7 6500U 2.5GHz    6              1TB HDD   
1302  Intel Celeron Dual Core N3050 1.6GHz    4            500GB HDD   

                               Gpu       OpSys  Weight_lb   $Price  \
0     Intel Iris Plus Graphics 640       macOS    3.02085  1527.25   
1           Intel HD Graphics 6000       macOS    2.95470  1024.79   
2            Intel HD Graphics 620       No OS    4.10130   655.50   
3               AMD Radeon Pro 455       macOS    4.03515  2892.69   
4     Intel Iris Plus Graphics 650       macOS    3.02085  2056.10   
...                            ...         ...        ...      ...   
1298         Intel HD Graphics 520  Windows 10    3.96900   727.32   
1299         Intel HD Graphics 520  Windows 10    2.86650  1708.86   
1300             Intel HD Graphics  Windows 10    3.30750   261.06   
1301            AMD Radeon R5 M330  Windows 10    4.82895   870.96   
1302             Intel HD Graphics  Windows 10    4.85100   420.66   

      TouchScreen  IPS          ppi    Cpu processor Processor_speed_GHz  
0               0    1  3018.873962  Intel Processor                 2.3  
1               0    0  1698.116604  Intel Processor                 1.8  
2               0    0  2202.907170  Intel Processor                 2.5  
3               0    1  3396.233208  Intel Processor                 2.7  
4               0    1  3018.873962  Intel Processor                 3.1  
...           ...  ...          ...              ...                 ...  
1298            1    1  2202.907170  Intel Processor                 2.5  
1299            1    1  3671.511950  Intel Processor                 2.5  
1300            0    0  1567.092850  Intel Processor                 1.6  
1301            0    0  1567.092850  Intel Processor                 2.5  
1302            0    0  1567.092850  Intel Processor                 1.6  

[1303 rows x 16 columns]

In [35]:
laptop_data.drop(['Cpu'], axis=1, inplace=True)
laptop_data.head()

Company      Product   TypeName  Inches  Ram               Memory  \
0   Apple  MacBook Pro  Ultrabook    13.3    8            128GB SSD   
1   Apple  Macbook Air  Ultrabook    13.3    8  128GB Flash Storage   
2      HP       250 G6   Notebook    15.6    8            256GB SSD   
3   Apple  MacBook Pro  Ultrabook    15.4   16            512GB SSD   
4   Apple  MacBook Pro  Ultrabook    13.3    8            256GB SSD   

                            Gpu  OpSys  Weight_lb   $Price  TouchScreen  IPS  \
0  Intel Iris Plus Graphics 640  macOS    3.02085  1527.25            0    1   
1        Intel HD Graphics 6000  macOS    2.95470  1024.79            0    0   
2         Intel HD Graphics 620  No OS    4.10130   655.50            0    0   
3            AMD Radeon Pro 455  macOS    4.03515  2892.69            0    1   
4  Intel Iris Plus Graphics 650  macOS    3.02085  2056.10            0    1   

           ppi    Cpu processor Processor_speed_GHz  
0  3018.873962  Intel Processor                 2.3  
1  1698.116604  Intel Processor                 1.8  
2  2202.907170  Intel Processor                 2.5  
3  3396.233208  Intel Processor                 2.7  
4  3018.873962  Intel Processor                 3.1

Momory column contains type of the memory: SSD, HDD, Flash Storage and Hybrid. Also, it contains the data storage capacity in GB and TB.

# Fetching the information from 'Memory' column

In [36]:
laptop_data['Memory'].value_counts()

256GB SSD                        412
1TB HDD                          223
500GB HDD                        132
512GB SSD                        118
128GB SSD +  1TB HDD              94
128GB SSD                         76
256GB SSD +  1TB HDD              73
32GB Flash Storage                38
2TB HDD                           16
64GB Flash Storage                15
512GB SSD +  1TB HDD              14
1TB SSD                           14
256GB SSD +  2TB HDD              10
1.0TB Hybrid                       9
256GB Flash Storage                8
16GB Flash Storage                 7
32GB SSD                           6
180GB SSD                          5
128GB Flash Storage                4
512GB SSD +  2TB HDD               3
16GB SSD                           3
512GB Flash Storage                2
1TB SSD +  1TB HDD                 2
256GB SSD +  500GB HDD             2
128GB SSD +  2TB HDD               2
256GB SSD +  256GB SSD             2
512GB SSD +  256GB SSD             1
5

In [37]:
#we removed 0 after the decimal point
#we removed 'GB' in memory column, and replaced 'TB' with '000' (1TB = 1000GB)
#some cell contains more than 1 memory info, they have '+' sign between two piece of info. we split those cell and save in new_df2. 
laptop_data['Memory'] = laptop_data['Memory'].astype(str).replace('\.0', '')
laptop_data['Memory'] = laptop_data['Memory'].str.replace('GB', '')
laptop_data['Memory'] = laptop_data['Memory'].str.replace('TB', '000')
new_df2 = laptop_data['Memory'].str.split('+', expand=True)
new_df2.head()

0     1
0            128 SSD  None
1  128 Flash Storage  None
2            256 SSD  None
3            512 SSD  None
4            256 SSD  None

In [38]:
#we are joing 0th index column of new_df2 with our original dataframe, creating new column 'para1'
#we are joing 1st index column of new_df2 with our original dataframe, creating new column 'para2'
laptop_data['para1'] = new_df2[0]
laptop_data['para2'] = new_df2[1]

laptop_data

Company                              Product            TypeName  Inches  \
0      Apple                          MacBook Pro           Ultrabook    13.3   
1      Apple                          Macbook Air           Ultrabook    13.3   
2         HP                               250 G6            Notebook    15.6   
3      Apple                          MacBook Pro           Ultrabook    15.4   
4      Apple                          MacBook Pro           Ultrabook    13.3   
...      ...                                  ...                 ...     ...   
1298  Lenovo                       Yoga 500-14ISK  2 in 1 Convertible    14.0   
1299  Lenovo                       Yoga 900-13ISK  2 in 1 Convertible    13.3   
1300  Lenovo                   IdeaPad 100S-14IBR            Notebook    14.0   
1301      HP  15-AC110nv (i7-6500U/6GB/1TB/Radeon            Notebook    15.6   
1302    Asus  X553SA-XX031T (N3050/4GB/500GB/W10)            Notebook    15.6   

      Ram             Memory                           Gpu       OpSys  \
0       8            128 SSD  Intel Iris Plus Graphics 640       macOS   
1       8  128 Flash Storage        Intel HD Graphics 6000       macOS   
2       8            256 SSD         Intel HD Graphics 620       No OS   
3      16            512 SSD            AMD Radeon Pro 455       macOS   
4       8            256 SSD  Intel Iris Plus Graphics 650       macOS   
...   ...                ...                           ...         ...   
1298    4            128 SSD         Intel HD Graphics 520  Windows 10   
1299   16            512 SSD         Intel HD Graphics 520  Windows 10   
1300    2   64 Flash Storage             Intel HD Graphics  Windows 10   
1301    6           1000 HDD            AMD Radeon R5 M330  Windows 10   
1302    4            500 HDD             Intel HD Graphics  Windows 10   

      Weight_lb   $Price  TouchScreen  IPS          ppi    Cpu processor  \
0       3.02085  1527.25            0    1  3018.873962  Intel Processor   
1       2.95470  1024.79            0    0  1698.116604  Intel Processor   
2       4.10130   655.50            0    0  2202.907170  Intel Processor   
3       4.03515  2892.69            0    1  3396.233208  Intel Processor   
4       3.02085  2056.10            0    1  3018.873962  Intel Processor   
...         ...      ...          ...  ...          ...              ...   
1298    3.96900   727.32            1    1  2202.907170  Intel Processor   
1299    2.86650  1708.86            1    1  3671.511950  Intel Processor   
1300    3.30750   261.06            0    0  1567.092850  Intel Processor   
1301    4.82895   870.96            0    0  1567.092850  Intel Processor   
1302    4.85100   420.66            0    0  1567.092850  Intel Processor   

     Processor_speed_GHz              para1 para2  
0                    2.3            128 SSD  None  
1                    1.8  128 Flash Storage  None  
2                    2.5            256 SSD  None  
3                    2.7            512 SSD  None  
4                    3.1            256 SSD  None  
...                  ...                ...   ...  
1298                 2.5            128 SSD  None  
1299                 2.5            512 SSD  None  
1300                 1.6   64 Flash Storage  None  
1301                 2.5           1000 HDD  None  
1302                 1.6            500 HDD  None  

[1303 rows x 17 columns]

In [39]:
#We create four columns Layer1HDD, Layer1SSD, Layer1Hybrid, Layer1Flash_Storage.
#If para1 contains HDD, then 1 will store in Layer1HDD column; otherwise 0. We did same with all four column.
laptop_data['Layer1HDD'] = laptop_data['para1'].apply(lambda x: 1 if 'HDD' in x else 0)
laptop_data['Layer1SSD'] = laptop_data['para1'].apply(lambda x: 1 if 'SSD' in x else 0)
laptop_data['Layer1Hybrid'] = laptop_data['para1'].apply(lambda x: 1 if 'Hybrid' in x else 0)
laptop_data['Layer1Flash_Storage'] = laptop_data['para1'].apply(lambda x: 1 if 'Flash Storage' in x else 0)

#We create another four columns Layer2HDD, Layer2SSD, Layer2Hybrid, Layer2Flash_Storage.
#If para2 contains HDD, then 1 will store in Layer2HDD column; otherwise 0. We did same with all four column.
laptop_data['para2'].fillna("0", inplace = True)
laptop_data['Layer2HDD'] = laptop_data['para2'].apply(lambda x: 1 if 'HDD' in x else 0)
laptop_data['Layer2SSD'] = laptop_data['para2'].apply(lambda x: 1 if 'SSD' in x else 0)
laptop_data['Layer2Hybrid'] = laptop_data['para2'].apply(lambda x: 1 if 'Hybrid' in x else 0)
laptop_data['Layer2Flash_Storage'] = laptop_data['para2'].apply(lambda x: 1 if 'Flash Storage' in x else 0)

laptop_data.sample(10)

Company                              Product            TypeName  Inches  \
33     Dell                               XPS 13           Ultrabook    13.3   
560    Acer                        Spin SP111-31  2 in 1 Convertible    11.6   
835    Dell                         Alienware 15              Gaming    15.6   
103      HP                       Envy 13-ad009n           Ultrabook    13.3   
883  Lenovo                    IdeaPad 510-15ISK            Notebook    15.6   
12    Apple                          MacBook Pro           Ultrabook    15.4   
442  Lenovo                        ThinkPad X270           Ultrabook    12.5   
639    Dell                               XPS 15            Notebook    15.6   
176    Acer                       Aspire A315-51            Notebook    15.6   
681    Asus  L502NA-GO052T (N3350/4GB/128GB/W10)            Notebook    15.6   

     Ram               Memory                      Gpu       OpSys  Weight_lb  \
33    16              512 SSD   Intel UHD Graphics 620  Windows 10    2.64600   
560    4     32 Flash Storage    Intel HD Graphics 500  Windows 10    2.75625   
835   16  256 SSD +  1000 HDD  Nvidia GeForce GTX 1070  Windows 10    7.07805   
103    8              256 SSD     Nvidia GeForce MX150  Windows 10    3.04290   
883    8             1000 HDD     Nvidia GeForce 940MX  Windows 10    4.85100   
12    16              256 SSD       AMD Radeon Pro 555       macOS    4.03515   
442    8              256 SSD    Intel HD Graphics 620  Windows 10    2.99880   
639   16             1000 SSD  Nvidia GeForce GTX 1050  Windows 10    4.54230   
176    4              128 SSD    Intel HD Graphics 520  Windows 10    4.63050   
681    4              128 SSD    Intel HD Graphics 500  Windows 10    4.10130   

      $Price  ...             para1       para2  Layer1HDD Layer1SSD  \
33   2130.66  ...           512 SSD           0          0         1   
560   397.86  ...  32 Flash Storage           0          0         0   
835  3163.08  ...          256 SSD     1000 HDD          0         1   
103  1275.66  ...           256 SSD           0          0         1   
883   762.66  ...          1000 HDD           0          1         0   
12   2781.57  ...           256 SSD           0          0         1   
442  1629.06  ...           256 SSD           0          0         1   
639  2734.86  ...          1000 SSD           0          0         1   
176   511.86  ...           128 SSD           0          0         1   
681   477.66  ...           128 SSD           0          0         1   

    Layer1Hybrid Layer1Flash_Storage Layer2HDD  Layer2SSD  Layer2Hybrid  \
33             0                   0         0          0             0   
560            0                   1         0          0             0   
835            0                   0         1          0             0   
103            0                   0         0          0             0   
883            0                   0         0          0             0   
12             0                   0         0          0             0   
442            0                   0         0          0             0   
639            0                   0         0          0             0   
176            0                   0         0          0             0   
681            0                   0         0          0             0   

     Layer2Flash_Storage  
33                     0  
560                    0  
835                    0  
103                    0  
883                    0  
12                     0  
442                    0  
639                    0  
176                    0  
681                    0  

[10 rows x 25 columns]

In [40]:
def mem_type(text):
    if text.split()[-1] == 'SSD':
        return 'SSD'
    elif text.split() [-1] == 'HDD':
        return 'HHD'
    elif text.split() [-1] == 'Storage':
        return 'Flash Storage'
    elif text.split() [-1] == 'Hybrid':
        return 'Hybrid'
    else:
        return ''

In [41]:
laptop_data['layer1Memorytype'] = laptop_data['para1'].apply(mem_type)
laptop_data['layer2Memorytype'] = laptop_data['para2'].apply(mem_type)
laptop_data['Memory_type'] = laptop_data['layer1Memorytype'] + ' ' + laptop_data['layer2Memorytype']
laptop_data.drop(columns=['layer1Memorytype', 'layer2Memorytype'], inplace=True)
laptop_data.sample(10)

Company                Product            TypeName  Inches  Ram  \
48      Dell          Inspiron 3567            Notebook    15.6    4   
959     Acer  Chromebook C738T-C2EJ  2 in 1 Convertible    11.6    4   
336     Dell          Inspiron 7570           Ultrabook    15.6    8   
896     Dell          Inspiron 7567              Gaming    15.6    8   
1208    Acer         Aspire E5-774G            Notebook    17.3    8   
664      MSI         GP72VR Leopard              Gaming    17.3    8   
884     Dell          Inspiron 3552            Notebook    15.6    4   
541     Dell            Vostro 5468            Notebook    14.0    4   
1271  Lenovo         Yoga 900-13ISK  2 in 1 Convertible    13.3   16   
158     Dell          Inspiron 5579  2 in 1 Convertible    15.6    8   

                   Memory                        Gpu       OpSys  Weight_lb  \
48                256 SSD         AMD Radeon R5 M430  Windows 10    5.07150   
959      32 Flash Storage      Intel HD Graphics 400   Chrome OS    2.75625   
336   128 SSD +  1000 HDD       Nvidia GeForce 940MX  Windows 10    4.18950   
896              1000 HDD  Nvidia GeForce GTX 1050Ti       Linux    5.77710   
1208             1000 HDD       Nvidia GeForce 940MX  Windows 10    7.27650   
664   128 SSD +  1000 HDD    Nvidia GeForce GTX 1060  Windows 10    5.95350   
884               500 HDD          Intel HD Graphics  Windows 10    4.85100   
541               128 SSD      Intel HD Graphics 520  Windows 10    3.52800   
1271              512 SSD      Intel HD Graphics 520  Windows 10    2.86650   
158               256 SSD     Intel UHD Graphics 620  Windows 10    3.43980   

       $Price  ...       para2  Layer1HDD  Layer1SSD Layer1Hybrid  \
48     728.46  ...           0          0          1            0   
959    443.46  ...           0          0          0            0   
336   1195.86  ...    1000 HDD          0          1            0   
896   1059.06  ...           0          1          0            0   
1208   717.06  ...           0          1          0            0   
664   1694.92  ...    1000 HDD          0          1            0   
884    420.66  ...           0          1          0            0   
541    838.89  ...           0          0          1            0   
1271  1708.86  ...           0          0          1            0   
158   1195.86  ...           0          0          1            0   

     Layer1Flash_Storage Layer2HDD Layer2SSD  Layer2Hybrid  \
48                     0         0         0             0   
959                    1         0         0             0   
336                    0         1         0             0   
896                    0         0         0             0   
1208                   0         0         0             0   
664                    0         1         0             0   
884                    0         0         0             0   
541                    0         0         0             0   
1271                   0         0         0             0   
158                    0         0         0             0   

      Layer2Flash_Storage     Memory_type  
48                      0            SSD   
959                     0  Flash Storage   
336                     0         SSD HHD  
896                     0            HHD   
1208                    0            HHD   
664                     0         SSD HHD  
884                     0            HHD   
541                     0            SSD   
1271                    0            SSD   
158                     0            SSD   

[10 rows x 26 columns]

In [42]:
#We removed string from para1 and para2
laptop_data['para1'] = laptop_data['para1'].str.replace('HDD', '')
laptop_data['para1'] = laptop_data['para1'].str.replace('SSD', '')
laptop_data['para1'] = laptop_data['para1'].str.replace('Hybrid', '')
laptop_data['para1'] = laptop_data['para1'].str.replace('Flash Storage', '')

laptop_data['para2'] = laptop_data['para2'].str.replace('HDD', '')
laptop_data['para2'] = laptop_data['para2'].str.replace('SSD', '')
laptop_data['para2'] = laptop_data['para2'].str.replace('Hybrid', '')
laptop_data['para2'] = laptop_data['para2'].str.replace('Flash Storage', '')

#It gives error with this code and with int32 at the next line code.
#laptop_data['para1'] = round(laptop_data['para1'], 0)
#laptop_data['para2'] = round(laptop_data['para2'], 0)

laptop_data['para1'] = laptop_data['para1'].astype('float64')
laptop_data['para2'] = laptop_data['para2'].astype('float64')
laptop_data

Company                              Product            TypeName  Inches  \
0      Apple                          MacBook Pro           Ultrabook    13.3   
1      Apple                          Macbook Air           Ultrabook    13.3   
2         HP                               250 G6            Notebook    15.6   
3      Apple                          MacBook Pro           Ultrabook    15.4   
4      Apple                          MacBook Pro           Ultrabook    13.3   
...      ...                                  ...                 ...     ...   
1298  Lenovo                       Yoga 500-14ISK  2 in 1 Convertible    14.0   
1299  Lenovo                       Yoga 900-13ISK  2 in 1 Convertible    13.3   
1300  Lenovo                   IdeaPad 100S-14IBR            Notebook    14.0   
1301      HP  15-AC110nv (i7-6500U/6GB/1TB/Radeon            Notebook    15.6   
1302    Asus  X553SA-XX031T (N3050/4GB/500GB/W10)            Notebook    15.6   

      Ram             Memory                           Gpu       OpSys  \
0       8            128 SSD  Intel Iris Plus Graphics 640       macOS   
1       8  128 Flash Storage        Intel HD Graphics 6000       macOS   
2       8            256 SSD         Intel HD Graphics 620       No OS   
3      16            512 SSD            AMD Radeon Pro 455       macOS   
4       8            256 SSD  Intel Iris Plus Graphics 650       macOS   
...   ...                ...                           ...         ...   
1298    4            128 SSD         Intel HD Graphics 520  Windows 10   
1299   16            512 SSD         Intel HD Graphics 520  Windows 10   
1300    2   64 Flash Storage             Intel HD Graphics  Windows 10   
1301    6           1000 HDD            AMD Radeon R5 M330  Windows 10   
1302    4            500 HDD             Intel HD Graphics  Windows 10   

      Weight_lb   $Price  ...  para2  Layer1HDD  Layer1SSD Layer1Hybrid  \
0       3.02085  1527.25  ...    0.0          0          1            0   
1       2.95470  1024.79  ...    0.0          0          0            0   
2       4.10130   655.50  ...    0.0          0          1            0   
3       4.03515  2892.69  ...    0.0          0          1            0   
4       3.02085  2056.10  ...    0.0          0          1            0   
...         ...      ...  ...    ...        ...        ...          ...   
1298    3.96900   727.32  ...    0.0          0          1            0   
1299    2.86650  1708.86  ...    0.0          0          1            0   
1300    3.30750   261.06  ...    0.0          0          0            0   
1301    4.82895   870.96  ...    0.0          1          0            0   
1302    4.85100   420.66  ...    0.0          1          0            0   

     Layer1Flash_Storage  Layer2HDD  Layer2SSD  Layer2Hybrid  \
0                      0          0          0             0   
1                      1          0          0             0   
2                      0          0          0             0   
3                      0          0          0             0   
4                      0          0          0             0   
...                  ...        ...        ...           ...   
1298                   0          0          0             0   
1299                   0          0          0             0   
1300                   1          0          0             0   
1301                   0          0          0             0   
1302                   0          0          0             0   

      Layer2Flash_Storage     Memory_type  
0                       0            SSD   
1                       0  Flash Storage   
2                       0            SSD   
3                       0            SSD   
4                       0            SSD   
...                   ...             ...  
1298                    0            SSD   
1299                    0            SSD   
1300                    0  Flash Storage   
1301                    0        

In [43]:
#laptop_data['Total_memory_size'] = laptop_data['para1'] + laptop_data['para1']
#laptop_data['Total_memory_size'] = laptop_data['Total_memory_size'].astype('int')
#laptop_data['Total_memory_size']

In [44]:
#Now we have eight columns, with 1 and 0 values. Those shows the type of the memory.
#But we want the actual storage capacity. so we multiplied all 8 columns with para1 and para2.
laptop_data['HDD'] = laptop_data['para1']*laptop_data['Layer1HDD'] + laptop_data['para2']*laptop_data['Layer2HDD']
laptop_data['SSD'] = laptop_data['para1']*laptop_data['Layer1SSD'] + laptop_data['para2']*laptop_data['Layer2SSD']
laptop_data['Hybrid'] = laptop_data['para1']*laptop_data['Layer1Hybrid'] + laptop_data['para2']*laptop_data['Layer2Hybrid']
laptop_data['Flash_storage'] = laptop_data['para1']*laptop_data['Layer1Flash_Storage'] + laptop_data['para2']*laptop_data['Layer2Flash_Storage']

laptop_data['HDD'] = round(laptop_data['HDD'], 0)
laptop_data['SSD'] = round(laptop_data['SSD'], 0)
laptop_data['Hybrid'] = round(laptop_data['Hybrid'], 0)
laptop_data['Flash_storage'] = round(laptop_data['Flash_storage'], 0)

In [45]:
#Drop unnecessory columns
laptop_data.drop(columns=['Memory', 'para1', 'para2', 'Layer1HDD', 'Layer1SSD', 'Layer1Hybrid', 'Layer1Flash_Storage', 'Layer2HDD', 'Layer2SSD', 'Layer2Hybrid', 'Layer2Flash_Storage', 'Memory_type'], inplace=True)

Company            Product  TypeName  Inches  Ram  \
153      MSI         GE73VR 7RE    Gaming    17.3   16   
927       HP        Probook 650  Notebook    15.6    8   
1110      HP     EliteBook 1030  Notebook    13.3    8   
610   Lenovo       Thinkpad P51  Notebook    15.6   32   
106   Lenovo  Ideapad 310-15ISK  Notebook    15.6    4   

                          Gpu       OpSys  Weight_lb   $Price  TouchScreen  \
153   Nvidia GeForce GTX 1060  Windows 10    6.17400  2154.60            0   
927     Intel HD Graphics 520  Windows 10    5.09355  1328.10            0   
1110    Intel HD Graphics 515  Windows 10    2.55780  1936.86            1   
610      Nvidia Quadro M2200M  Windows 10    5.51250  5584.86            0   
106     Intel HD Graphics 520  Windows 10    4.07925   472.99            0   

      IPS         ppi    Cpu processor Processor_speed_GHz     HDD     SSD  \
153     0  2202.90717  Intel Processor                 2.8  1000.0   256.0   
927     0  2202.90717  Intel Processor                 2.3     0.0   256.0   
1110    0  3671.51195  Intel Processor                 1.1     0.0   256.0   
610     1  4405.81434  Intel Processor                 3.1     0.0  1000.0   
106     0  2202.90717  Intel Processor                   2  1000.0     0.0   

      Hybrid  Flash_storage  
153      0.0            0.0  
927      0.0            0.0  
1110     0.0            0.0  
610      0.0            0.0  
106      0.0            0.0

In [47]:
laptop_data['HDD'] = laptop_data['HDD'].astype('int32')
laptop_data['SSD'] = laptop_data['SSD'].astype('int32')
laptop_data['Hybrid'] = laptop_data['Hybrid'].astype('int32')
laptop_data['Flash_storage'] = laptop_data['Flash_storage'].astype('int32')
laptop_data.head()

Company      Product   TypeName  Inches  Ram                           Gpu  \
0   Apple  MacBook Pro  Ultrabook    13.3    8  Intel Iris Plus Graphics 640   
1   Apple  Macbook Air  Ultrabook    13.3    8        Intel HD Graphics 6000   
2      HP       250 G6   Notebook    15.6    8         Intel HD Graphics 620   
3   Apple  MacBook Pro  Ultrabook    15.4   16            AMD Radeon Pro 455   
4   Apple  MacBook Pro  Ultrabook    13.3    8  Intel Iris Plus Graphics 650   

   OpSys  Weight_lb   $Price  TouchScreen  IPS          ppi    Cpu processor  \
0  macOS    3.02085  1527.25            0    1  3018.873962  Intel Processor   
1  macOS    2.95470  1024.79            0    0  1698.116604  Intel Processor   
2  No OS    4.10130   655.50            0    0  2202.907170  Intel Processor   
3  macOS    4.03515  2892.69            0    1  3396.233208  Intel Processor   
4  macOS    3.02085  2056.10            0    1  3018.873962  Intel Processor   

  Processor_speed_GHz  HDD  SSD  Hybrid  Flash_storage  
0                 2.3    0  128       0              0  
1                 1.8    0    0       0            128  
2                 2.5    0  256       0              0  
3                 2.7    0  512       0              0  
4                 3.1    0  256       0              0

In [48]:
laptop_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Company              1303 non-null   object 
 1   Product              1303 non-null   object 
 2   TypeName             1303 non-null   object 
 3   Inches               1303 non-null   float64
 4   Ram                  1303 non-null   int32  
 5   Gpu                  1303 non-null   object 
 6   OpSys                1303 non-null   object 
 7   Weight_lb            1303 non-null   float32
 8   $Price               1303 non-null   float64
 9   TouchScreen          1303 non-null   int64  
 10  IPS                  1303 non-null   int64  
 11  ppi                  1303 non-null   float64
 12  Cpu processor        1303 non-null   object 
 13  Processor_speed_GHz  1303 non-null   object 
 14  HDD                  1303 non-null   int32  
 15  SSD                  1303 non-null   i

# Fetching the information from 'Gpu' column

In [49]:
laptop_data['Gpu'].value_counts()

Intel HD Graphics 620      281
Intel HD Graphics 520      185
Intel UHD Graphics 620      68
Nvidia GeForce GTX 1050     66
Nvidia GeForce GTX 1060     48
                          ... 
AMD Radeon R5 520            1
AMD Radeon R7                1
Intel HD Graphics 540        1
AMD Radeon 540               1
ARM Mali T860 MP4            1
Name: Gpu, Length: 110, dtype: int64

In [50]:
def gpu(text):
    if text[0] == 'AMD':
        return 'Good'
    elif text[0] == 'MSI':
        return 'Good'
    elif text[0] == 'Nvidia':
        return 'Good'
    else:
        return 'Average'

In [51]:
laptop_data['Gpu'] = laptop_data['Gpu'].apply(gpu)
laptop_data.sample(10)

Company                                Product            TypeName  \
366     Dell                          Inspiron 5770            Notebook   
560     Acer                          Spin SP111-31  2 in 1 Convertible   
1170      HP                                 250 G5            Notebook   
740     Dell                          Inspiron 7560            Notebook   
326     Acer                          Aspire E5-575            Notebook   
418     Dell                          Inspiron 3567            Notebook   
337       HP                          Elitebook 850            Notebook   
487     Dell                                 XPS 15            Notebook   
811      MSI                            GE72MVR 7RG              Gaming   
333   Lenovo  V310-15ISK (i3-6006U/4GB/1TB/FHD/W10)            Notebook   

      Inches  Ram      Gpu       OpSys  Weight_lb   $Price  TouchScreen  IPS  \
366     17.3    8  Average  Windows 10    6.17400  1236.90            0    0   
560     11.6    4  Average  Windows 10    2.75625   397.86            1    1   
1170    15.6    8  Average  Windows 10    4.32180   774.06            0    0   
740     15.6    8  Average  Windows 10    4.41000  1375.98            0    0   
326     15.6    6  Average  Windows 10    4.91715   625.86            0    0   
418     15.6    8  Average  Windows 10    4.71870   615.54            0    0   
337     15.6    8  Average  Windows 10    4.05720  1304.16            0    0   
487     15.6   16  Average  Windows 10    4.54230  2164.86            0    0   
811     17.3   16  Average  Windows 10    6.39450  2753.10            0    0   
333     15.6    4  Average  Windows 10    4.07925   511.86            0    0   

             ppi    Cpu processor Processor_speed_GHz   HDD  SSD  Hybrid  \
366   2202.90717  Intel Processor                 1.6  1000  128       0   
560   2202.90717  Intel Processor                 2.0     0    0       0   
1170  2202.90717  Intel Processor                 2.5     0  256       0   
740   2202.90717  Intel Processor                 2.7  1000  128       0   
326   1567.09285  Intel Processor                 2.5  1000    0       0   
418   1567.09285  Intel Processor                 2.5  1000    0       0   
337   2202.90717  Intel Processor                 2.5     0  256       0   
487   2202.90717  Intel Processor                 2.8     0  512       0   
811   2202.90717  Intel Processor                 2.8  1000  512       0   
333   2202.90717  Intel Processor                   2  1000    0       0   

      Flash_storage  
366               0  
560              32  
1170              0  
740               0  
326               0  
418               0  
337               0  
487               0  
811               0  
333               0

When you’re looking for a Graphics Card,  The best graphics cards are MSI, AMD, and Nvidia. So I considered these cards as good and others as average.

# Fetching the information from 'OpSys' column

In [52]:
laptop_data['OpSys'].value_counts()

Windows 10      1072
No OS             66
Linux             62
Windows 7         45
Chrome OS         27
macOS             13
Mac OS X           8
Windows 10 S       8
Android            2
Name: OpSys, dtype: int64

In [53]:
def operating_system(text):
    if text.split()[0] == 'Windows':
        return 'Windows'
    elif text.split()[0] == 'Linux':
        return 'Linux'
    elif text.split()[0] == 'Mac' and 'macOs':
        return 'Mac'
    elif text.split()[0] == 'Android':
        return 'Android'
    elif text.split()[0] == 'Chrome':
        return 'Chrome'
    else:
        return 'No OS'

In [54]:
laptop_data['OpSys'] = laptop_data['OpSys'].apply(operating_system)
laptop_data.sample()

Company                            Product  TypeName  Inches  Ram  \
134      HP  15-BS078nr (i7-7500U/8GB/1TB/W10)  Notebook    15.6    8   

         Gpu    OpSys  Weight_lb  $Price  TouchScreen  IPS         ppi  \
134  Average  Windows    4.52025  681.72            0    0  1567.09285   

       Cpu processor Processor_speed_GHz   HDD  SSD  Hybrid  Flash_storage  
134  Intel Processor                 2.7  1000    0       0              0

and lastly, I assigned idividual operating system

In [55]:
#product name doesn't helpful so dropping it
laptop_data['Product'].value_counts()

XPS 13                                     30
Inspiron 3567                              29
250 G6                                     21
Legion Y520-15IKBN                         19
Vostro 3568                                19
                                           ..
15-bw007nv (A10-9620P/6GB/128GB/Radeon      1
SmartBook 130                               1
A541NA-GO342 (N3350/4GB/500GB/Linux)        1
17-X047na (i3-6006U/8GB/1TB/W10)            1
V330-15IKB (i5-8250U/4GB/500GB/FHD/W10)     1
Name: Product, Length: 618, dtype: int64

In [56]:
laptop_data.drop(['Product'], axis=1, inplace=True)
laptop_data

Company            TypeName  Inches  Ram      Gpu    OpSys  Weight_lb  \
0      Apple           Ultrabook    13.3    8  Average    No OS    3.02085   
1      Apple           Ultrabook    13.3    8  Average    No OS    2.95470   
2         HP            Notebook    15.6    8  Average    No OS    4.10130   
3      Apple           Ultrabook    15.4   16  Average    No OS    4.03515   
4      Apple           Ultrabook    13.3    8  Average    No OS    3.02085   
...      ...                 ...     ...  ...      ...      ...        ...   
1298  Lenovo  2 in 1 Convertible    14.0    4  Average  Windows    3.96900   
1299  Lenovo  2 in 1 Convertible    13.3   16  Average  Windows    2.86650   
1300  Lenovo            Notebook    14.0    2  Average  Windows    3.30750   
1301      HP            Notebook    15.6    6  Average  Windows    4.82895   
1302    Asus            Notebook    15.6    4  Average  Windows    4.85100   

       $Price  TouchScreen  IPS          ppi    Cpu processor  \
0     1527.25            0    1  3018.873962  Intel Processor   
1     1024.79            0    0  1698.116604  Intel Processor   
2      655.50            0    0  2202.907170  Intel Processor   
3     2892.69            0    1  3396.233208  Intel Processor   
4     2056.10            0    1  3018.873962  Intel Processor   
...       ...          ...  ...          ...              ...   
1298   727.32            1    1  2202.907170  Intel Processor   
1299  1708.86            1    1  3671.511950  Intel Processor   
1300   261.06            0    0  1567.092850  Intel Processor   
1301   870.96            0    0  1567.092850  Intel Processor   
1302   420.66            0    0  1567.092850  Intel Processor   

     Processor_speed_GHz   HDD  SSD  Hybrid  Flash_storage  
0                    2.3     0  128       0              0  
1                    1.8     0    0       0            128  
2                    2.5     0  256       0              0  
3                    2.7     0  512       0              0  
4                    3.1     0  256       0              0  
...                  ...   ...  ...     ...            ...  
1298                 2.5     0  128       0              0  
1299                 2.5     0  512       0              0  
1300                 1.6     0    0       0             64  
1301                 2.5  1000    0       0              0  
1302                 1.6   500    0       0              0  

[1303 rows x 17 columns]

In [58]:
clean_data = laptop_data

In [59]:
%store clean_data

Stored 'clean_data' (DataFrame)
